js/packages/openinference-instrumentation-llama-index/src/utils.ts

```diff
@@ -161,25 +228,90 @@ export function patchRetrieveMethod(
+  tracer: Tracer,
 ) {
-  const docs: Attributes = {};
-  output.forEach((document, index) => {
-    const { node, score } = document;
+  return function patchedQueryEmbedding(
+    this: unknown,
+    ...args: Parameters<QueryEmbeddingMethod>
+  ) {
+    const span = tracer.startSpan(`embedding`, {
+      kind: SpanKind.INTERNAL,
+      attributes: {
+        [SemanticConventions.OPENINFERENCE_SPAN_KIND]:
+          OpenInferenceSpanKind.EMBEDDING,
+      },
+    });
 
-    if (node instanceof TextNode) {
-      const nodeId = node.id_;
-      const nodeText = node.getContent();
-      const nodeMetadata = node.metadata;
+    const execContext = getExecContext(span);
 
-      const prefix = `${SemanticConventions.RETRIEVAL_DOCUMENTS}.${index}`;
-      docs[`${prefix}.${SemanticConventions.DOCUMENT_ID}`] = nodeId;
-      docs[`${prefix}.${SemanticConventions.DOCUMENT_SCORE}`] = score;
-      docs[`${prefix}.${SemanticConventions.DOCUMENT_CONTENT}`] = nodeText;
-      docs[`${prefix}.${SemanticConventions.DOCUMENT_METADATA}`] =
-        safelyJSONStringify(nodeMetadata) ?? undefined;
-    }
-  });
-  return docs;
+    const execPromise = safeExecuteInTheMiddle<
+      ReturnType<QueryEmbeddingMethod>
+    >(
+      () => {
+        return context.with(execContext, () => {
+          return original.apply(this, args);
+        });
+      },
+      (error) => handleError(span, error),
+    );
+
+    // Model ID/name is a property found on the class and not in args
+    // Extract from class and set as attribute
+    span.setAttributes({
+      [SemanticConventions.EMBEDDING_MODEL_NAME]: getModelName(this),
+    });
+
+    const wrappedPromise = execPromise.then((result) => {
+      span.setAttributes(queryEmbeddingAttributes(args[0], result));
+      span.end();
+      return result;
+    });
+    return context.bind(execContext, wrappedPromise);
+  };
+}
+
+export function patchTextEmbeddings(
+  original: TextEmbeddingsMethod,
+  tracer: Tracer,
+) {
+  return function patched(
+    this: unknown,
+    ...args: Parameters<TextEmbeddingsMethod>
+  ) {
+    const span = tracer.startSpan(`embedding`, {
+      kind: SpanKind.INTERNAL,
+      attributes: {
+        [SemanticConventions.OPENINFERENCE_SPAN_KIND]:
+          OpenInferenceSpanKind.EMBEDDING,
+      },
+    });
+
+    const execContext = getExecContext(span);
+
+    const execPromise = safeExecuteInTheMiddle<
+      ReturnType<TextEmbeddingsMethod>
+    >(
+      () => {
+        return context.with(execContext, () => {
+          return original.apply(this, args);
+        });
+      },
+      (error) => handleError(span, error),
+    );
+
+    // Model ID/name is a property found on the class and not in args
+    // Extract from class and set as attribute
+    span.setAttributes({
+      [SemanticConventions.EMBEDDING_MODEL_NAME]: getModelName(this),
+    });
+
+    const wrappedPromise = execPromise.then((result) => {
+      span.setAttributes(textEmbeddingsAttributes(args[0], result));
```

Comment:
you don't have to overly abstract things, and this might be a case where you don't think it makes sense, but it looks like this function "textEmbedding..." is the only difference between text patch and query patch. looking above it looks like "textEmb..." and "queryEmb..." functions can be merged into one, you can just use the text embed one and pass in an array withonly one element in the query case